src/compiler/c.rs

```diff
@@ -267,21 +268,25 @@ pub const CACHED_ENV_VARS : &'static [&'static str] = &[
 ];
 
 /// Compute the hash key of `compiler` compiling `preprocessor_output` with `args`.
-pub fn hash_key(compiler_digest: &str, arguments: &str, env_vars: &[(OsString, OsString)],
+pub fn hash_key(compiler_digest: &str,
+                arguments: &[OsString],
+                env_vars: &[(OsString, OsString)],
                 preprocessor_output: &[u8]) -> String
 {
     // If you change any of the inputs to the hash, you should change `CACHE_VERSION`.
```

Comment:
Nah I think in practice this is going to end up disturbing the hashes a bit. The motivation of `HashToSha1` was to avoid transmuting `OsStr` to bytes (which while valid today isn't something we'd like to commit to in libstd). 

Is changing the hash though a problem? I could back that out if it's best left for another time.